ethcore/light/src/on_demand/mod.rs

```diff
@@ -54,13 +54,13 @@ struct Peer {
 }
 
 impl Peer {
-	// whether this peer can fulfill the
+	// whether this peer can fulfill the necessary capabilities
 	fn can_fulfill(&self, c: &Capabilities) -> bool {
 		let caps = &self.capabilities;
 
-		caps.serve_headers == c.serve_headers &&
-			caps.serve_chain_since >= c.serve_chain_since &&
-			caps.serve_state_since >= c.serve_chain_since
+		caps.serve_headers >= c.serve_headers &&
+			(c.serve_chain_since.is_none() || caps.serve_chain_since <= c.serve_chain_since) &&
+			(c.serve_state_since.is_none() || caps.serve_state_since <= c.serve_state_since)
```

Comment:
if the request doesn't require any state or chain serving, then it doesn't matter what the peer capabilities are. If the request does require chain or state serving, then the block number this peer supports beyond should be `<=` the request required block number. but this should also guard against `caps.serve_since.is_none()`. I will rewrite for clarity